// openweather-api-node/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AIR_POLLUTION_PATH = exports.ONECALL_PATH = exports.FORECAST_PATH = exports.WEATHER_PATH = exports.DATA_3_PATH = exports.DATA_PATH = exports.GEO_PATH = exports.API_ENDPOINT = exports.SUP_UNITS = exports.SUP_LANGS = void 0;
exports.SUP_LANGS = [
    "af",
    "al",
    "ar",
    "az",
    "bg",
    "ca",
    "cz",
    "da",
    "de",
    "el",
    "en",
    "eu",
    "fa",
    "fi",
    "fr",
    "gl",
    "he",
    "hi",
    "hr",
    "hu",
    "id",
    "it",
    "ja",
    "kr",
    "la",
    "lt",
    "mk",
    "no",
    "nl",
    "pl",
    "pt",
    "pt_br",
    "ro",
    "ru",
    "sv",
    "se",
    "sk",
    "sl",
    "sp",
    "es",
    "sr",
    "th",
    "tr",
    "ua",
    "uk",
    "vi",
    "zh_cn",
    "zh_tw",
    "zu",
];
exports.SUP_UNITS = ["standard", "metric", "imperial"];
exports.API_ENDPOINT = "https://api.openweathermap.org/";
exports.GEO_PATH = "geo/1.0/";
exports.DATA_PATH = "data/2.5";
exports.DATA_3_PATH = "data/3.0";
exports.WEATHER_PATH = exports.DATA_PATH + "/weather";
exports.FORECAST_PATH = exports.DATA_PATH + "/forecast";
exports.ONECALL_PATH = exports.DATA_3_PATH + "/onecall";
exports.AIR_POLLUTION_PATH = exports.DATA_PATH + "/air_pollution";
